src/layers/rotary_embedding_qwen.cpp

```diff
@@ -327,6 +327,94 @@ void QwenRotaryEmbedding::forward(
+    int dim = this->inv_freq_size * 2;
+    REQUIRES(dim == qkShape[3], "Incorrect shape, this dimention is not the head size.");
+
+    const int batchSize = qkShape[0];
+    const int seqLen = qkShape[1];
+    const int qHeads = qkShape[2];
+    const int kHeads = qkShape[4];
+    const int maxSeqLength = qkShape[5];
+    const int pastKeyLength = qkShape[6];
+    const int heads = std::max(qHeads, kHeads);
+    const int half = this->inv_freq_size;
+    int kv_len = seqLen + pastKeyLength;
+    REQUIRES(kv_len < maxSupportedSeqLength, "process seq length must less than 32768.");
+
+    float new_base = getNewBaseValue(kv_len, maxSeqLength);
+    if (std::abs(new_base - this->base) > 1e-5) {
+        this->base = new_base;
+
+        auto it = embCosSin.find(new_base);
+        if (it == embCosSin.end()) {
+            float *inv_freq = (float *)malloc(this->inv_freq_size * sizeof(float));
+#pragma omp parallel for
+            for (size_t i = 0; i < this->inv_freq_size; i++) {
+                inv_freq[i] = 1.0 / pow(base, float(i * 2) / dim);
+            }
+            QwenCalEmb(inv_freq, new_base, embCosSin);
+            free(inv_freq);
+        }
+
+        auto &value = embCosSin[new_base];
+        cur_emb_cos = std::get<0>(value);
+        cur_emb_sin = std::get<1>(value);
+    }
+
+    float *q_scale = logn + pastKeyLength;
+#pragma omp parallel for collapse(3)
+    for (int head = 0; head < heads; ++head) {
+        for (int bs = 0; bs < batchSize; ++bs) {
+            for (int seq = 0; seq < seqLen; ++seq) {
+                int pos = positionIds[seq];
+                float *pcos = cur_emb_cos + pos * dim;
+                float *psin = cur_emb_sin + pos * dim;
+
+                float16_t *q = query + bs * seqLen * qStride + seq * qStride + head * dim;
+                float16_t *k = key + bs * seqLen * kStride + seq * kStride + head * dim;
+
+                __m512 pScale = _mm512_set1_ps(q_scale[seq]);
+
+                // Process chunks of 16 elements at a time
+                for (int i = 0; i < half; i += 16) {
+                    int remain = half - i;
+                    __mmask16 mask = (remain >= 16 ? 0xffff : (1 << remain) - 1);
+
+                    __m512 pCosVec = _mm512_maskz_loadu_ps(mask, &pcos[i]);
+                    __m512 pCosHalfVec = _mm512_maskz_loadu_ps(mask, &pcos[i + half]);
+                    __m512 pSinVec = _mm512_maskz_loadu_ps(mask, &psin[i]);
+                    __m512 pSinHalfVec = _mm512_maskz_loadu_ps(mask, &psin[i + half]);
+
+                    // Compute something like:
+                    // q[i] = ( q[i] * pcos[i] - q[i + half] * psin[i] ) * logN;
+                    // q[i + half] = ( q[i + half] * pcos[i + half] + q[i] * psin[i + half] ) * logN;
+                    if (head < qHeads) {
+                        __m512 qVec = float16_t::cvt_fp16_to_fp32(_mm256_maskz_loadu_epi16(mask, &q[i]));
```

Comment:
consider to use xft::load_avx512 to make the code simpler?

Reply:
Done